[WIP] Add (optional) extended dashboards in ScyllaDBMonitoring and drop the unmanaged setup

rzetelskik opened the pull request:
**Description of your changes:** WIP

**Which issue is resolved by this Pull Request:**
Resolves #1155

tnozicka commented on assets/monitoring/prometheus/v1/scylladb.servicemonitor.yaml:
```diff
@@ -5,7 +5,7 @@ metadata:
   selector:
     {{- .endpointsSelector | toYAML | nindent 4 }}
-  jobLabel: scylla/cluster
+  jobLabel: app.kubernetes.io/name
```
this needs to stay - this is the label from the service so in the end, if you name your scyllacluster scylla, it will make sure there is a job=scylla label used by some dashboards

rzetelskik replied:
All of scylla-monitoring dashboards , including the serverless one, have "job" label hardcoded to either `scylla` or `scylla-manager`. This change breaks multiple metrics. What do you propose?

rzetelskik replied:
Discussed internally. Temporarily fixed job labels in metrics used by dashboards. This is supposed to be fixed in upstream dashboards in long term.

tnozicka commented on pkg/api/scylla/v1alpha1/types_monitoring.go:
```diff
@@ -162,6 +162,16 @@ type ScyllaDBMonitoringSpec struct {
+	// "Platform" means the monitoring is deployed in the ScyllaDB cluster owner's platform. A full monitoring setup
+	// designed for cluster administrators is deployed.
+	// "SaaS" means the monitoring is deployed for a SaaS offering. A simplified monitoring setup is deployed.
+	// +kubebuilder:default:="Platform"
```
sounds like breaking change

rzetelskik replied:
Would you rather have no default then and make it required or use "SaaS" as default?

rzetelskik replied:
@tnozicka ping

tnozicka replied:
in v1alpha1 it should behave the same. with "alpha" you can break when introducing v1alpha2

it needs to be set in a way that it keeps behaving the same way (if you have deployed it already it shouldn't change your dashboards type when updating the operator+crd), which would be default=SaaS, at least for v1alpha1

rzetelskik replied:
Right. I changed the default to `SaaS`, but explicitly set `Platform` in out examples. I believe it would make sense to change the default to `Platform` with the next API version though.

tnozicka replied:
I feel the same, my point was purely about API compatibility guarantees

tnozicka commented on pkg/api/scylla/v1alpha1/types_monitoring.go:
```diff
@@ -162,6 +162,16 @@ type ScyllaDBMonitoringSpec struct {
+	// designed for cluster administrators is deployed.
+	// "SaaS" means the monitoring is deployed for a SaaS offering. A simplified monitoring setup is deployed.
+	// +kubebuilder:default:="Platform"
+	// +kubebuilder:validation:Enum="";"SaaS";"Platform"
```
why ""?

rzetelskik replied:
I thought it was necessary to pass validation with implicit value. Removed, thanks.

tnozicka reviewed (commented):
a few comments while scanning the approach

tnozicka commented on assets/monitoring/prometheus/v1/scylladbmanager.servicemonitor.yaml:
```diff
@@ -4,14 +4,14 @@ metadata:
-  jobLabel: "app.kubernetes.io/name"
+  jobLabel: app.kubernetes.io/name
   selector:
     matchLabels:
```
i think this would need to be configurable

tnozicka commented on assets/monitoring/prometheus/v1/scylladbmanager.servicemonitor.yaml:
```diff
@@ -4,14 +4,14 @@ metadata:
+  jobLabel: app.kubernetes.io/name
   selector:
     matchLabels:
       app.kubernetes.io/name: scylla-manager
```
i think this would need to be configurable

rzetelskik replied:
Sure, that was only a WIP. Dropping the manager support until we have an automated manager setup as agreed.

tnozicka commented on pkg/api/scylla/v1alpha1/types_monitoring.go:
```diff
@@ -162,6 +162,16 @@ type ScyllaDBMonitoringSpec struct {
 	// components hold additional config for the monitoring components in use.
 	Components *Components `json:"components"`
+
+	// type determines what platform service model the monitoring is deployed for. Defaults to SaaS. Valid options
+	// are Platform and SaaS.
+	// "Platform" means the monitoring is deployed in the ScyllaDB cluster owner's platform. A full monitoring setup
+	// designed for cluster administrators is deployed.
+	// "SaaS" means the monitoring is deployed for a SaaS offering. A simplified monitoring setup is deployed.
+	// +kubebuilder:default:="SaaS"
+	// +kubebuilder:validation:Enum="SaaS";"Platform"
```
```suggestion
	// type determines the deployment type of the monitoring setup. Can be "Platform" or "Saas". 
	// "Platform" sets up a detailed ScyllaDB monitoring, including the infrastructure.
	// "Saas" deploys monitoring setup focusing on the ScyllaDB service.
	// Defaults to "SaaS".
	// +kubebuilder:validation:Enum="SaaS";"Platform"
	// +kubebuilder:default:="SaaS"
```

tnozicka replied:
(i think it's better to first introduce the values and then say what the default is)

rzetelskik replied:
It sort of copied it from core Service object, but I don't mind either.

tnozicka commented on pkg/controller/scylladbmonitoring/sync_grafana.go:
```diff
@@ -140,9 +140,18 @@ func makeGrafanaConfigs(sm *scyllav1alpha1.ScyllaDBMonitoring) (*corev1.ConfigMa
+			"scyllaDBMonitoringName": sm.Name,
+		})
+	default:
+		return nil, "", fmt.Errorf("unkown type: %s", sm.Spec.Type)
```
nit: `switch t:=`
also `unknown monitoring type`

tnozicka commented on docs/source/monitoring.md:
```diff
@@ -1,76 +1,130 @@
+
+## Deploy managed monitoring
+
+**Note**: as of v1.8, ScyllaDBMonitoring is in alpha state and is susceptible to backwards incompatible changes.
```
```suggestion
**Note**: as of v1.8, ScyllaDBMonitoring is experimental. The API is currently at v1alpha1 version and may change in the future versions.
```

tnozicka commented on docs/source/monitoring.md:
```diff
@@ -1,76 +1,130 @@
+
+### Requirements
+
+Before you can set up your ScyllaDB Monitoring, you must install the Scylla Operator in your Kubernetes cluster 
```
```suggestion
Before you can set up your ScyllaDB Monitoring, you have to have Scylla Operator already installed in your Kubernetes cluster
```

tnozicka commented on docs/source/monitoring.md:
```diff
@@ -1,76 +1,130 @@
+### Requirements
+
+Before you can set up your ScyllaDB Monitoring, you must install the Scylla Operator in your Kubernetes cluster 
+and have an instance of ScyllaCluster running. For details see:
```
i don't think that's technically true

tnozicka replied:
You can find more information on how to deploy Scylla Operator here:

tnozicka commented on docs/source/monitoring.md:
```diff
@@ -1,76 +1,130 @@
+* [Deploying Scylla on a Kubernetes Cluster](generic.md)
+* [Deploying Scylla stack using Helm Charts](helm.md)
+
+The managed monitoring setup also requires HAProxy Ingress and Prometheus Operator deployed in your Kubernetes cluster.
```
it's not required by the monitoring itself, it's just that the example uses it

tnozicka commented on docs/source/monitoring.md:
```diff
@@ -1,76 +1,130 @@
+You can deploy them using the provided third party examples.
+
+#### Deploy Prometheus Operator
+Deploy Prometheus Operator using kubectl:
```
if your cluster already has a prometheus operator deployed you should skip this step

rzetelskik replied:
I've put a message above saying you can skip these steps if you already have them deployed in your cluster.

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,130 @@
+
+##### Verify Prometheus Operator status
+```console
+$ kubectl -n prometheus-operator get deployments.apps/prometheus-operator
````
kubectl rollout status

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,130 @@
+```
+
+#### Deploy HAProxy Ingress
+Deploy HAProxy Ingress using kubectl:
````
same skip here

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,130 @@
+
+##### Verify HAProxy Ingress status
+```console
+$ kubectl -n haproxy-ingress get deployments.apps/haproxy-ingress
````
rollout status

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,130 @@
+```
+
+Before you begin, update the `endpointSelector` in `examples/monitoring/v1alpha1/scylladbmonitoring.yaml` with a label 
+matching your ScyllaCluster instance.
````
instance name.

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,130 @@
+
+Deploy the monitoring setup using kubectl:
+```console
+$ kubectl apply --server-side -f examples/monitoring/v1alpha1/scylladbmonitoring.yaml
````
```suggestion
$ kubectl apply --server-side -f ./examples/monitoring/v1alpha1/scylladbmonitoring.yaml
```

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,130 @@
+$ kubectl apply --server-side -f examples/monitoring/v1alpha1/scylladbmonitoring.yaml
+```
+
+The Operator should pick up the object's creation and create all the necessary resources for you.
````
```suggestion
Scylla Operator will notice the new ScyllaDBMonitoring object and create and manage all the necessary resources for you.
```

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,130 @@
+
+#### Verify ScyllaDBMonitoring status
+```console
+$ kubectl get ScyllaDBMonitoring example
````
wait for conditions

rzetelskik replied:
There are multiple so I had to use a workaround

rzetelskik replied:
Also asked on https://github.com/kubernetes/kubernetes/issues/95759 if I can give fixing this a try

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,130 @@
+example-grafana-admin-credentials   Opaque   2      68s
+```
+
+### Access Grafana locally
````
I think this should be a variant when you don't want to use/install haproxy, or expose it to the world

rzetelskik replied:
Do you suggest removing it or adding it as an alternative?

tnozicka replied:
I think we'd primarily describe a way to use the ingress, this would be an alternative, if needed

tnozicka commented on examples/monitoring/v1alpha1/scylladbmonitoring.yaml:
```diff
@@ -26,4 +26,3 @@ spec:
             - test-grafana.test.svc.cluster.local
             annotations:
               haproxy-ingress.github.io/ssl-passthrough: "true"
-
```
why?

zimnx commented on pkg/api/scylla/v1alpha1/types_monitoring.go:
```diff
@@ -162,6 +162,15 @@ type ScyllaDBMonitoringSpec struct {
+	// "Platform" sets up a detailed ScyllaDB monitoring, including the infrastructure.
+	// "Saas" deploys monitoring setup focusing on the ScyllaDB service.
+	// Defaults to "SaaS".
+	// +kubebuilder:validation:Enum="SaaS";"Platform"
```
Since enumeration values are part of the API, shouldn't these be declared as consts in api package instead in `naming`?

rzetelskik replied:
good point, changed

zimnx commented on pkg/api/scylla/v1alpha1/types_monitoring.go:
```diff
@@ -162,6 +162,15 @@ type ScyllaDBMonitoringSpec struct {
+	// +kubebuilder:validation:Enum="SaaS";"Platform"
+	// +kubebuilder:default:="SaaS"
+	// +optional
+	Type string `json:"type,omitempty"`
```
ScyllaDBMonitoringType instead of string

tnozicka replied:
now i see it has both the new and old changes which needs cleanup

rzetelskik commented:
@zimnx @tnozicka thanks, I believe I've addressed all of your comments

zimnx reviewed (dismissed):
lgtm, thanks

tnozicka commented:
> Note: we decided not to include ScyllaDB Manager dashboards until we start managing ScyllaDB Manager deployments with Operator.

Given the plan for manager is changing, I'd use something more general, like:
`Note: We have decided not to include ScyllaDB Manager dashboards yet until we figure out the how we'll integrate manager functionality in a better way.`

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,148 @@
+example-grafana-admin-credentials   Opaque   2      68s
+```
+
+### Access Grafana
````
I feel that this article is too focused on how to hack it with minikube and someone trying to set this up on a real cluster / domain would be confused.

I think the generic steps are:
- create ScyllaDBMonitoring with a domain (`test-grafana.test.svc.cluster.local` is fine)
- [real domain / cname to ingress controller] curl
- [fake domain / port-forward / not a cname to ingress controller] curl + `--resolve`

rzetelskik replied:
> - [real domain / cname to ingress controller] curl
> - [fake domain / port-forward / not a cname to ingress controller] curl + --resolve

I don't understand these lines. Could you elaborate on what these options in the brackets mean? Do you mean that there should be two sections demonstrating all of these options, verifying them with either curl or curl --resolve? Are you also suggesting not using minikube to demonstrate it?

tnozicka replied:
I think that could be a joint section saying if your scylldbmonitoring domain isn't resolvable (to the ingress controller IP accessible from outside of the cluster), then you have to add the --resolve flag with ingress controller IP or use `/etc/hosts` alias. The ingress controller IP (accessible from outside of the cluster) can be constructed in a few ways - as service externalIP, nodeIP +  service nodePort, or 127.0.0.1+kubectl port-forward to ingress controller svc, ...

tnozicka commented on pkg/api/scylla/v1alpha1/types_monitoring.go:
```diff
@@ -170,7 +177,7 @@ type ScyllaDBMonitoringSpec struct {
 	// +kubebuilder:default:="SaaS"
 	// +optional
-	Type string `json:"type,omitempty"`
+	Type ScyllaDBMonitoringType `json:"type,omitempty"`
```
ok, it's here but there needs to be just one commit  touching this enum

tnozicka commented on pkg/naming/constants.go:
```diff
@@ -148,10 +148,3 @@ type ProtocolDNSLabel string
 	CQLProtocolDNSLabel = "cql"
 )
-
-type ScyllaDBMonitoringType string
```
you need to clean your commits

rzetelskik commented on assets/monitoring/grafana/v1alpha1/dashboards-saas.cm.yaml:
```diff
@@ -4602,15 +4602,15 @@ data:
                         },
                         "datasource": "prometheus",
-                        "definition": "query_result(count(up{job=\"scylla\"}) by (dc))",
+                        "definition": "query_result(count(up{job=~\"$cluster|$^\"}) by (dc))",
```
Sent https://github.com/scylladb/scylla-operator/pull/1199

rzetelskik commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,148 @@
+$ kubectl get secrets/example-grafana-admin-credentials
+NAME                                TYPE     DATA   AGE
+example-grafana-admin-credentials   Opaque   2      68s
+```
````
Yeah it doesn't really make sense, I left it there to demonstrate how to access the admin credentials. Changed.

vponomaryov commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,142 @@
+deployment "haproxy-ingress" successfully rolled out
+```
+
+Before you begin, update the `endpointSelector` in `examples/monitoring/v1alpha1/scylladbmonitoring.yaml` with a label 
````
The code uses plural form of the first word -> `endpointsSelector`.

tnozicka commented on docs/source/monitoring.md:
```diff
@@ -1,76 +1,142 @@
+$ kubectl wait --for='condition=Available=True' scylladbmonitorings.scylla.scylladb.com/example
+scylladbmonitoring.scylla.scylladb.com/example condition met
+
+$ kubectl wait --for='condition=Progressing=False' scylladbmonitorings.scylla.scylladb.com/example
```
with this raciness, you should check progressing first and available last. it can be still available for a while when it starts progressing and when it rolls out you won't check it again

rzetelskik replied:
good point, updated

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,234 @@
+X-Frame-Options: deny
+X-Xss-Protection: 1; mode=block
+Date: Fri, 07 Apr 2023 14:32:09 GMT
+```
````
I've tried to simplify this and also some of the templates. Please verify the instructions.

### Accessing Grafana

For accessing Grafana service from outside the Kubernetes cluster we'd recommend using an Ingress but there are many other ways. What matters when using the Ingress is to direct your packets to the ingress controller service/pods and have the correct TLS SNI field set by the caller when reaching out the service, so it is routed properly, and it can successfully validate the grafana serving certificate. Obviously this is easier when you are using a real DNS domain that resolves to your Ingress controller IP address but most of the client/tools allow to set the SNI field manually.

#### Prerequisites

For accessing Grafana, you first need to collect the serving CA and the credentials.

```console
$ GRAFANA_SERVING_CERT="$( kubectl -n scylla get secret/example-grafana-serving-ca --template '{{ index .data "tls.crt" }}' | base64 -d )"
$ GRAFANA_USER="$( kubectl -n scylla get secret/example-grafana-admin-credentials --template '{{ index .data "username" }}' | base64 -d )"
$ GRAFANA_PASSWORD="$( kubectl -n scylla get secret/example-grafana-admin-credentials --template '{{ index .data "password" }}' | base64 -d )"
```

#### Connecting through Ingress using a resolvable domain

In case of production cluster someone has already set up the Ingress controller and appropriate DNS records. Often there is a generic wildcard record like `*.app.mydomain` already pointing to the ingress controller external IP. For custom service domains, these are usually a CNAME pointing the Ingress controller A record.

Note: The ScyllaDBMonitoring example creates an Ingress object with `test-grafana.test.svc.cluster.local` DNS domain - you need to adjust this to your domain. Here it is represented as `example-grafana.apps.mydomain`.

Note: To test a resolvable domain from your machine without creating DNS records, you can adjust `/etc/hosts` or similar.

```console
$ curl --fail -s -o /dev/null -w '%{http_code}' -L --cacert <( echo "${GRAFANA_SERVING_CERT}" ) "https://example-grafana.apps.mydomain" --user "${GRAFANA_USER}:${GRAFANA_PASSWORD}"
200
```

#### Connecting through Ingress using an unresolvable domain

To connect to an ingress without a resolvable domain you first need to find out your Ingress controller IP that can be resolved externally. Again there are more ways to do it, not limited to the few examples shown here.

Unless stated otherwise, we will assume your ingress is running on port 443.
```console
$ INGRESS_PORT=443
```

##### Variants
###### Ingress ExternalIP

When you are running in a real cluster there is usually a cloud LB or a bare metal alternative providing you with an externally reachable IP.

```console
$ INGRESS_IP="$( kubectl -n=haproxy-ingress get service/haproxy-ingress --template='{{ ( index .loadBalancer.ingress 0 ).ip }}' )"
```

###### Ingress NodePort

NodePort is a bit of a less convenient but it's available in dev clusters as well.

```console
# FIXME: the template needs to prefer ExternalIP, if set. 
$ INGRESS_IP="$( kubectl get nodes --template='{{ range ( index .items 0 ).status.addresses }}{{ if eq .type "InternalIP" }}{{ .address }}{{ end }}{{ end }}' )"
$ INGRESS_PORT="$( kubectl -n=haproxy-ingress get services/haproxy-ingress --template='{{ range .spec.ports }}{{ if eq .port 443 }}{{ .nodePort }}{{ end }}{{ end }}' )"
```

###### Ingress Port Forwarding

Port forwarding may be useful, if your nodes are not reachable through a public IP, nor your Ingress, or in general for development purposes. 

```console
$ kubectl -n=haproxy-ingress port-forward services/haproxy-ingress 5000:443 >/dev/null &
$ INGRESS_IP="localhost"
$ INGRESS_PORT="5000"
```

##### Connection

```console
$ curl --fail -s -o /dev/null -w '%{http_code}' -L --cacert <( echo "${GRAFANA_SERVING_CERT}" ) "https://test-grafana.test.svc.cluster.local"  --resolve "test-grafana.test.svc.cluster.local:${INGRESS_PORT}:${INGRESS_IP}" --user "${GRAFANA_USER}:${GRAFANA_PASSWORD}"
200
```

rzetelskik replied:
I guess we had a different perspective on how this should look like but this looks good to me - I'll vet the commands and fix the typos and other errors.

rzetelskik commented on pkg/controller/scylladbmonitoring/sync_grafana.go:
```diff
@@ -140,9 +140,18 @@ func makeGrafanaConfigs(sm *scyllav1alpha1.ScyllaDBMonitoring) (*corev1.ConfigMa
-	return grafanav1alpha1assets.GrafanaDashboardsConfigMapTemplate.RenderObject(map[string]any{
-		"scyllaDBMonitoringName": sm.Name,
-	})
+	switch t := *sm.Spec.Type; t {
```
It has a default value, so it should never happen. I've added a sanity check though.

rzetelskik replied:
Good point! I assumed it would never happen because of the default value, but I haven't thought of the version skew.

rzetelskik commented on assets/monitoring/grafana/v1alpha1/registry.go:
```diff
@@ -36,9 +36,13 @@ var (
 
-	//go:embed "dashboards.cm.yaml"
-	grafanaDashboardsConfigMapTemplateString string
-	GrafanaDashboardsConfigMapTemplate       = ParseObjectTemplateOrDie[*corev1.ConfigMap]("grafana-dashboard-cm", grafanaDashboardsConfigMapTemplateString)
```
It doesn't get pruned but it get updated, yes.

tnozicka replied:
oh, right, the name is still the same, thx

rzetelskik commented:
@tnozicka thanks for the review. As discussed internally I've dropped the `port-forward` part and will add it once I've investigated the issue. I've changed the wording of the last section here and there and fixed some errors so you might want to double check.

tnozicka commented on pkg/controller/scylladbmonitoring/sync_grafana.go:
```diff
@@ -140,9 +140,24 @@ func makeGrafanaConfigs(sm *scyllav1alpha1.ScyllaDBMonitoring) (*corev1.ConfigMa
+	if sm.Spec.Type == nil {
+		// This should have a default value, but it can be nil due to a version skew.
+		scyllaDBMonitoringTypeSAAS := scyllav1alpha1.ScyllaDBMonitoringTypeSAAS
+		sm.Spec.Type = &scyllaDBMonitoringTypeSAAS
```
this will break the informers, you can't write to the shared cache. pls make a variable for the type to switch on

rzetelskik replied:
Fixed. Btw should I add a test for `makeGrafanaDashboards`?

tnozicka replied:
I don't mind wither way but I'd not hold this over it

rzetelskik replied:
Let's go ahead then

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,180 @@
-   curl -L https://github.com/scylladb/scylla-monitoring/tarball/branch-4.0 | tar -xzf - -C scylla-monitoring --strip-components=1
-   ```
+For accessing Grafana service from outside the Kubernetes cluster we recommend using an Ingress, although there are many other ways to do so.
+When using Ingress, what matters is to direct your packets to the ingress controller Service/Pods and have the correct TLS SNI field set by the caller when reaching out to the service, so it is routed properly, and it can successfully validate the grafana serving certificate. 
````
```suggestion
When using Ingress, what matters is to direct your packets to the ingress controller Service/Pods and have the correct TLS SNI field set by the caller when reaching out to the service, so it is routed properly, and your client can successfully validate the grafana serving certificate.
```

tnozicka commented on docs/source/monitoring.md:
````diff
@@ -1,76 +1,180 @@
+NodePort is slightly less convenient, but it's available in development clusters as well.
+
+```console
+$ INGRESS_IP="$( kubectl get nodes --template='{{ $internal_ip := "" }}{{ $external_ip := "" }}{{ range ( index .items 0 ).status.addresses }}{{ if eq .type "InternalIP" }}{{ $internal_ip = .address }}{{ else if eq .type "ExternalIP" }}{{ $external_ip = .address }}{{ end }}{{ end }}{{ if $external_ip }}{{ $external_ip }}{{ else }}{{ $internal_ip }}{{ end }}' )"
````
Shouldn't we prefer the `ExternalIP` first and `InternalIP` as a fallback as it work only if you are in the same network like with your VM but not for say a cloud? The other way around both should work with a fallback.

tnozicka replied:
you are right, I've misread while scrolling

tnozicka commented on pkg/controller/scylladbmonitoring/sync_grafana.go:
```diff
@@ -140,9 +140,25 @@ func makeGrafanaConfigs(sm *scyllav1alpha1.ScyllaDBMonitoring) (*corev1.ConfigMa
-	return grafanav1alpha1assets.GrafanaDashboardsConfigMapTemplate.RenderObject(map[string]any{
-		"scyllaDBMonitoringName": sm.Name,
-	})
+	smType := sm.Spec.Type
+	if smType == nil {
+		// This should have a default value, but it can be nil due to a version skew.
+		scyllaDBMonitoringTypeSAAS := scyllav1alpha1.ScyllaDBMonitoringTypeSAAS
+		smType = &scyllaDBMonitoringTypeSAAS
+	}
+
+	switch t := *smType; t {
```
there is no need for this to be a pointer
```suggestion
	t := scyllav1alpha1.ScyllaDBMonitoringTypeSAAS
	if sm.Spec.Type != nil {
	    t = *sm.Spec.Type
	}
	
	
	switch t {
```

tnozicka reviewed (approved):
/approve
/lgtm
thanks!

mykaul commented:
> /approve /lgtm thanks!

Hallelujah!
